src/Functions/CMakeLists.txt

```diff
@@ -51,71 +53,81 @@ option(STRIP_DEBUG_SYMBOLS_FUNCTIONS "Do not generate debugger info for ClickHou
 else()
     message(STATUS "Generating debugger info for ClickHouse functions")
 endif()
 
 if (TARGET ch_contrib::icu)
-    target_link_libraries (clickhouse_functions PRIVATE ch_contrib::icu)
+    list (APPEND PRIVATE_LIBS ch_contrib::icu)
 endif ()
 
 if (TARGET ch_contrib::fastops)
-    target_link_libraries (clickhouse_functions PRIVATE ch_contrib::fastops)
+    list (APPEND PRIVATE_LIBS ch_contrib::fastops)
 endif ()
 
 if (ENABLE_EXAMPLES)
     add_subdirectory(examples)
 endif ()
 
 if (TARGET ch_contrib::llvm)
-    target_link_libraries(clickhouse_functions PRIVATE ch_contrib::llvm)
+    list (APPEND PRIVATE_LIBS ch_contrib::llvm)
 endif ()
 
 if (TARGET ch_contrib::base-x)
-    target_link_libraries(clickhouse_functions PRIVATE ch_contrib::base-x)
+    list (APPEND PRIVATE_LIBS ch_contrib::base-x)
 endif()
 
 if (TARGET ch_contrib::base64)
-    target_link_libraries(clickhouse_functions PRIVATE ch_contrib::base64)
+    list (APPEND PRIVATE_LIBS ch_contrib::base64)
 endif()
 
-target_link_libraries(clickhouse_functions PRIVATE ch_contrib::lz4)
+list (APPEND PRIVATE_LIBS ch_contrib::lz4)
 
 if (ENABLE_NLP)
-    target_link_libraries(clickhouse_functions PRIVATE ch_contrib::cld2)
+    list (APPEND PRIVATE_LIBS ch_contrib::cld2)
 endif()
 
 if (TARGET ch_contrib::h3)
-    target_link_libraries (clickhouse_functions PRIVATE ch_contrib::h3)
+    list (APPEND PRIVATE_LIBS ch_contrib::h3)
 endif()
 
 if (TARGET ch_contrib::vectorscan)
-    target_link_libraries(clickhouse_functions PRIVATE ch_contrib::vectorscan)
+    list (APPEND PRIVATE_LIBS ch_contrib::vectorscan)
 endif()
 
 if (TARGET ch_contrib::simdjson)
-    target_link_libraries(clickhouse_functions PRIVATE ch_contrib::simdjson)
+    list (APPEND PRIVATE_LIBS ch_contrib::simdjson)
 endif()
 
 if (TARGET ch_contrib::rapidjson)
-    target_link_libraries(clickhouse_functions PRIVATE ch_contrib::rapidjson)
+    list (APPEND PRIVATE_LIBS ch_contrib::rapidjson)
 endif()
 
 add_subdirectory(GatherUtils)
-target_link_libraries(clickhouse_functions PRIVATE clickhouse_functions_gatherutils)
+list (APPEND PRIVATE_LIBS clickhouse_functions_gatherutils)
 
 add_subdirectory(URL)
-target_link_libraries(clickhouse_functions PRIVATE clickhouse_functions_url)
+list (APPEND OBJECT_LIBS $<TARGET_OBJECTS:clickhouse_functions_url>)
 
 add_subdirectory(array)
-target_link_libraries(clickhouse_functions PRIVATE clickhouse_functions_array)
+list (APPEND OBJECT_LIBS $<TARGET_OBJECTS:clickhouse_functions_array>)
 
 add_subdirectory(JSONPath)
+list (APPEND PRIVATE_LIBS clickhouse_functions_jsonpath)
 
 # Signed integer overflow on user-provided data inside boost::geometry - ignore.
 set_source_files_properties("pointInPolygon.cpp" PROPERTIES COMPILE_FLAGS -fno-sanitize=signed-integer-overflow)
 
+# __start and __stop reserved identifier is used for loop through all the registered functions
+set_source_files_properties("registerFunctions.cpp" PROPERTIES COMPILE_FLAGS -Wno-reserved-identifier)
```

Comment:
> @liyinsg Thanks for the big effort!
> 
> Your main goal is to speed up linking if a single function changed, right? To be honest, I find the new mechanism rather hard-to-understand and brittle (e.g. placing functions into a special section of the binary, usage of object libraries which are a somewhat exotic feature of CMake).
> 
> Instead, what about splitting the function library into one sub-library per function group, e.g. clickhouse_functions_datetime.a, clickhouse_functions_string.a, etc.? This will still significantly reduce link times but not increase complexity as much.
> 
> > to eliminate the step to add and call an extern function in the registerFunctions.cpp
> 
> The current registration mechanism is a bit verbose (as one may expect for a hierarchical tree of functions) but not really error-prone, IMHO. Worst case is to forget registering a function in which case it simply won't be available in SQL.

@rschu1ze Thanks a lot for the review. My main goal is to eliminate the step to touch multiple files when adding a new function.  For example before we need to write something like below (`registerFunctionsGeo.cpp`):
```cpp
#if USE_H3
    registerFunctionGeoToH3(factory);
...
#endif

#if USE_S2_GEOMETRY
    registerFunctionGeoToS2(factory);

```
With the patch this step can be removed, since `geoToH3.cpp` is already enclosed by `USE_H3`.

I will try to fix other review comments later.

Reply:
> There is precedent (l. 120), but in general, it is nicer and more direct if compiler warnings are suppressed via "pragma diagnostic push/ignored/pop" directly in the source/header file.

The compile can still continue if the compiler doesn't support the flag if we put the flag in the cmake file, but not in the source file (`error: unknown pragma ignored` will be reported).